### DISCUSSION TITLE
Web UI for lightnovel-crawler

### DISCUSSION THREAD

**Discussion by hoangphuc05** [lncrawl/lightnovel-crawler#1740] [Show and tell]
# Adding a Graphical User Interface to lncrawl
I think ln crawler is pretty awesome, and adding some UI to help operate the software would be pretty nice.

I am aware that there's an attempt to make a web server (?), but seems like it has been dropped?

While there are multiple ways to create a native GUI on Python, I find most of them hard to use, and even harder to customize. Instead of using native UI frameworks like qt or Tkinter, I think implementing a web UI that can be served from the software would be easier for new people to contribute and extend functionality.

### My idea of the website tech stack/form:
* Website will be accessible through something like `localhost:9000`
* The website is made by ReactJS, styling with TailwindCSS.
* The website can communicate with the lncrawl by API (Flask, FastAPI?)

### (Expected) Features
1. We can start with something simple as operating basic function of the program from the website.
2. Document API the website use so that other program can talk to lncrawl as a client
3. Slowly add all functions available in the command line to the website
4. Future: Allow users to read downloaded light novels directly on the web UI
5. ...

# Here are some quick prototypes that I made for what the website would look like

## Download novel directly by URL
![image](https://user-images.githubusercontent.com/15669865/198503004-b3b8486c-88f3-4c00-9b6c-d131e56520a2.png)

## Search novel by title, and specifying source:
![image](https://user-images.githubusercontent.com/15669865/198503088-35b78446-692c-4f10-990d-5d502e37edc5.png)
### Choose a novel from found result:
![image](https://user-images.githubusercontent.com/15669865/198503144-add0d2bd-c35c-4156-bd0c-2589cdd69918.png)
### Options before finally download:
![image](https://user-images.githubusercontent.com/15669865/198503202-dbd422af-5ecf-4d72-adbd-bb9fc6e85168.png)

### Link to the prototype figma: https://www.figma.com/file/hCWw0CLRUjWgLOBfbGnCcZ/lncrawl?node-id=0%3A1
Disclaimer: I'm not a professional designer, and thus I copied a lot of wireframe/UI toolkit for this prototype

## Thought?
Would love to hear some feedback, idea, and opinion on both the overall idea and the look and feel of the prototype. 
This is something that I can work on too, but would love to hear opinions before spending time making it!

**Comment by PolarisFx** [comment]
I like the idea, but I'd also prefer some library management features. In my mind I see a new *Arr script like ReadArr with not only the ability to search, but to manage your library and be able to perform automated updates as chapters are released.

**Comment by jere344** [comment]
I'm the one making a website using lightnovel-crawler (I think you were talking about me ?) and am still actively working on it. I just started hosting it today on http://lncrawler.monster/

But my goals are really different, so I believe having both can be great. Your web UI concept is way better and user friendly for downloading novel while I only focused on the library and public access aspect, so there is no downloading parameters (no way to download ebooks from it currently for exemple).

To make your project work and be easily upgradeable I believe the easiest would be a layer above the console instead of directly making a bot.

Perhaps I can try to help on the backend stuff if there is a demand for a web UI downloader.

**Comment by dipu-bd** [comment]
If I can support a web UI, I can also publish a website collecting novels from various sources. Let's think about the original authors and translators for a bit in this situation. I do not want to push their buttons too hard with a very easy to use interface that everyone can use for grabbing all the novels everywhere.

**Reply by dipu-bd** [reply]
That being said, I have a plan to make a web server bot with `flask`. You can do all sort of operations through a standard set of API interface. If you want to make a web UI it necessary to make the API design first. There is an old open issue for this: #625

**Reply by hoangphuc05** [reply]
Good point. 
My (personal) argument is that we already passed the point where we crawl and grab novels from other sources. Making the tool easier to use doesn't really change/worsen the "moral" of our tool. Says if something is "wrong" then making it harder to approach doesn't make it "right". 
I think as long as we focus on developing this tool for personal uses instead of full fledge web server then we should be fine

**Reply by hoangphuc05** [reply]
And of course I always respect the decision of not having this as a feature.

**Comment by novelheart** [comment]
That being said, I have a plan to make a web server bot with [novelheart](https://novelheart.com/)
